src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp

```diff
@@ -577,46 +609,69 @@ void ParquetBlockInputFormat::initializeIfNeeded()
         return std::min(std::max(preferred_num_rows, MIN_ROW_NUM), static_cast<size_t>(format_settings.parquet.max_block_size));
     };
 
-    std::unique_ptr<ParquetBloomFilterCondition> parquet_bloom_filter_condition;
-
     std::unordered_set<std::size_t> filtering_columns;
 
     if (format_settings.parquet.bloom_filter_push_down && key_condition)
     {
         bf_reader = parquet::BloomFilterReader::Make(arrow_file, metadata, bf_reader_properties, nullptr);
 
-        const auto parquet_conditions = keyConditionRPNToParquetBloomFilterCondition(
+        const auto parquet_conditions = keyConditionRPNToParquetRPN(
             key_condition->getRPN(),
             index_mapping,
             metadata->RowGroup(0));
-        parquet_bloom_filter_condition = std::make_unique<ParquetBloomFilterCondition>(parquet_conditions, getPort().getHeader());
 
-        filtering_columns = parquet_bloom_filter_condition->getFilteringColumnKeys();
+        filtering_columns = getBloomFilterFilteringColumnKeys(parquet_conditions);
     }
 
-    for (int row_group = 0; row_group < num_row_groups; ++row_group)
+    auto skip_row_group_based_on_filters = [&](int row_group)
     {
-        if (skip_row_groups.contains(row_group))
-            continue;
+        if (!format_settings.parquet.filter_push_down && !format_settings.parquet.bloom_filter_push_down)
+        {
+            return false;
+        }
+
+        KeyCondition::RPN possibly_modified_rpn = key_condition->getRPN();
+        KeyCondition::ColumnIndexToBloomFilter column_index_to_bloom_filter;
+
+        const auto & header = getPort().getHeader();
+
+        std::vector<Range> hyperrectangle(header.columns(), Range::createWholeUniverse());
 
-        if (parquet_bloom_filter_condition)
+        if (format_settings.parquet.filter_push_down)
         {
-            const auto column_index_to_bf = buildColumnIndexToBF(*bf_reader, row_group, index_mapping, filtering_columns);
+            hyperrectangle = getHyperrectangleForRowGroup(*metadata, row_group, header, format_settings);getHyperrectangleForRowGroup(*metadata, row_group, getPort().getHeader(), format_settings);
+        }
 
-            if (!parquet_bloom_filter_condition->mayBeTrueOnRowGroup(column_index_to_bf))
-            {
-                continue;
-            }
+        if (format_settings.parquet.bloom_filter_push_down)
+        {
+            possibly_modified_rpn = keyConditionRPNToParquetRPN(key_condition->getRPN(),
```

Comment:
Save the result from the `keyConditionRPNToParquetRPN` call above instead of calling it again for each row group.